tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py

```diff
@@ -0,0 +1,621 @@
+        Utils.initialize_model_parallel(1, 1)
+
+        model, optimizer = setup_model_and_optimizer(
+            seed=2,
+            tp=1,
+            pp=1,
+            bf16=True,
+            dist_opt=False,
+            initialize_fn=initialize_gpt_model,
+            optimizer='dist_muon',
+        )
+
+        # Check if optimizer is ChainedOptimizer (expected for standard setup)
+        if isinstance(optimizer, ChainedOptimizer):
+            total_params = sum(
+                len(group['params'])
+                for opt in optimizer.chained_optimizers
+                for group in opt.param_groups
+            )
+            assert total_params > 0, "No parameters found in optimizer"
+
+    @pytest.mark.parametrize('tp_pp', [(1, 2), (2, 1), (2, 2)])
+    def test_broadcast_params(self, tp_pp):
+        """Test that parameter broadcasting works correctly across DP ranks."""
+        Utils.initialize_model_parallel(*tp_pp)
+
+        model, optimizer = setup_model_and_optimizer(
+            seed=2,
+            tp=tp_pp[0],
+            pp=tp_pp[1],
+            bf16=True,
+            dist_opt=False,
+            initialize_fn=initialize_gpt_model,
+            optimizer='dist_muon',
+        )
+
+        # If this is a LayerWiseDistributedOptimizer, test broadcast
+        if isinstance(optimizer, LayerWiseDistributedOptimizer):
+            # Store original param values
+            original_params = {}
+            for name, param in model[0].named_parameters():
+                original_params[name] = param.data.clone()
+
+            # Call broadcast (should be idempotent if no updates)
+            optimizer.broadcast_params()
+
+            # Check params are unchanged after broadcast without step
+            for name, param in model[0].named_parameters():
+                assert torch.allclose(param.data, original_params[name])
+
+    @pytest.mark.parametrize('tp_pp', [(2, 2), (2, 4), (4, 2)])
+    @pytest.mark.parametrize('bf16', [True, False])
+    def test_layer_wise_optimizer_save_load(self, tmp_path_dist_ckpt, tp_pp, bf16):
+        """Test save/load of LayerWiseDistributedOptimizer checkpoints."""
+        Utils.initialize_model_parallel(*tp_pp)
+
+        with TempNamedDir(
+            tmp_path_dist_ckpt / 'test_layer_wise_optimizer_A', sync=True
+        ) as ckpt_dir_A:
+            with TempNamedDir(
+                tmp_path_dist_ckpt / 'test_layer_wise_optimizer_B', sync=True
+            ) as ckpt_dir_B:
+                # Create model and optimizer A
+                model_A, optimizer_A = setup_model_and_optimizer(
+                    seed=2,
+                    tp=tp_pp[0],
+                    pp=tp_pp[1],
+                    bf16=bf16,
+                    dist_opt=False,
+                    initialize_fn=initialize_gpt_model,
+                    optimizer='dist_muon',
+                )
+
+                # Save checkpoint A
+                model_sharded_sd_A = model_A[0].sharded_state_dict()
```

Comment:
Can you use `generate_state_dict` at `training/checkpointing.py` instead? 
You can pass arguments to `False` or `none` if you want to exclude rng or opt_param_scheduler
This will make some `non-shardable` objects, which will be saved in `common.pt`

Reply:
I think the test below for common_state_dict serves this purpose

Reply:
Then we can resolve this conversation?